File: analyzer.py
# This is a sample Python script.

# Press Shift+F10 to execute it or replace it with your code.
# Press Double Shift to search everywhere for classes, files, tool windows, actions, and settings.
import time
from stringdata import get_data
def linear_search(container,element):
    for i in range(len(container)):
        if container[i]==element:
            return i
    return -1

def binary_search(container,element):
    #nonstanard implementation but helps eliminate off by 1 errors
    power=1
    power_counter=1
    while(power<len(container)):
        power*=2
        power_counter+=1
    ans=0
    for i in range(power_counter,-1,-1):
        if ans+(1<<i)<len(container) and container[ans+(1<<i)]<=element:
            ans+=(1<<i)
    #print(ans)
    #print(container[ans])
    if container[ans]==element:
        return ans
    else:
        return -1

# Press the green button in the gutter to run the script.
if __name__ == '__main__':
    data=get_data()
    string = "not_here"
    start_time=time.time()
    print("linear search, string",string,"found at",linear_search(data,string))
    end_time = time.time()
    print(f'Took {end_time - start_time:.15f} seconds')
    start_time = time.time()
    print("binary search, string", string, "found at", binary_search(data, string))
    end_time = time.time()
    print(f'Took {end_time - start_time:.15f} seconds')
    string = "mzzzz"
    start_time=time.time()
    print("linear search, string",string,"found at",linear_search(data,string))
    end_time = time.time()
    print(f'Took {end_time - start_time:.15f} seconds')
    start_time = time.time()
    print("binary search, string", string, "found at", binary_search(data, string))
    end_time = time.time()
    print(f'Took {end_time - start_time:.15f} seconds')
    string = "aaaaa"
    start_time=time.time()
    print("linear search, string",string,"found at",linear_search(data,string))
    end_time = time.time()
    print(f'Took {end_time - start_time:.15f} seconds')
    start_time = time.time()
    print("binary search, string", string, "found at", binary_search(data, string))
    end_time = time.time()
    print(f'Took {end_time - start_time:.15f} seconds')


# See PyCharm help at https://www.jetbrains.com/help/pycharm/

File: analyzer.cpp
#include <bits/stdc++.h>
#include "stringdata.h"
using namespace std;

int linear_search(vector<string>& container, string element){
	for(int i=0;i<container.size();i++){
		if(container[i]==element){
			return i;
		}
	}
	return -1;
}
int binary_search(vector<string>& container, string element){
	int power=1;
	int power_counter=1;
	while(power<container.size()){
		power*=2;
		power_counter+=1;
	}
	int ans=0;
	for(int i=power_counter;i>=0;i--){
		if(ans+(1<<i)<container.size() && container[ans+(1<<i)]<=element){
			ans+=(1<<i);
		}
	}
	if(container[ans]==element){
		return ans;
	}
	else{
		return -1;
	}
}

int main()
{
    vector<string> data=getStringData();
    string s="not_here";
    long long st=systemTimeNanoseconds();
    cout<<"linear search, string "<<s<<" found at "<<linear_search(data,s)<<endl;
    long long et=systemTimeNanoseconds();
    cout<<"took "<<et-st<<" nanoseconds"<<endl;
    st=systemTimeNanoseconds();
    cout<<"binary search, string "<<s<<" found at "<<binary_search(data,s)<<endl;
    et=systemTimeNanoseconds();
    cout<<"took "<<et-st<<" nanoseconds"<<endl;

    s="mzzzz";
    st=systemTimeNanoseconds();
    cout<<"linear search, string "<<s<<" found at "<<linear_search(data,s)<<endl;
    et=systemTimeNanoseconds();
    cout<<"took "<<et-st<<" nanoseconds"<<endl;
    st=systemTimeNanoseconds();
    cout<<"binary search, string "<<s<<" found at "<<binary_search(data,s)<<endl;
    et=systemTimeNanoseconds();
    cout<<"took "<<et-st<<" nanoseconds"<<endl;

    s="aaaaa";
    st=systemTimeNanoseconds();
    cout<<"linear search, string "<<s<<" found at "<<linear_search(data,s)<<endl;
    et=systemTimeNanoseconds();
    cout<<"took "<<et-st<<" nanoseconds"<<endl;
    st=systemTimeNanoseconds();
    cout<<"binary search, string "<<s<<" found at "<<binary_search(data,s)<<endl;
    et=systemTimeNanoseconds();
    cout<<"took "<<et-st<<" nanoseconds"<<endl;

    return 0;
}